File: app/src/main/java/com/emerjbl/ultra8/Chip8Input.java
package com.emerjbl.ultra8;

public class Chip8Input {

    boolean[] keys = new boolean[16];
    public boolean reset;
    public boolean pause;
    public boolean hurry = false;

    public boolean keyPressed(int x) {
        return keys[x];
    }

    public synchronized void reset() {
        reset = true;
        notify();
    }

    public void clear() {
        reset = false;
    }

    public void pause() {
        pause = true;
    }

    public void resume() {
        notify();
        pause = false;
    }

    public synchronized void setKey(int x) {
        keys[x] = true;
        notify();
    }

    public void resetKey(int x) {
        keys[x] = false;
    }

    public int checkForPress() {
        for (int i = 15; i >= 0; i--) {
            if (keys[i]) {
                return i;
            }
        }
        return -1;
    }


}

File: app/src/main/java/com/emerjbl/ultra8/Chip8Input.kt
package com.emerjbl.ultra8

class Chip8Input {
    var keys: BooleanArray = BooleanArray(16)

    fun keyPressed(x: Int): Boolean {
        return keys[x]
    }

    private val monitor = Any()

    @Synchronized
    fun setKey(x: Int) {
        keys[x] = true
        (this as Object).notify()
    }

    fun resetKey(x: Int) {
        keys[x] = false
    }

    fun awaitPress(): Int {
        val pressed = checkForPress()
        if (pressed >= 0) {
            return pressed
        }

        try {
            (monitor as Object).wait()
        } catch (ex: InterruptedException) {
            return -1
        }
        return checkForPress()
    }

    fun checkForPress(): Int {
        for (i in 15 downTo 0) {
            if (keys[i]) {
                return i
            }
        }
        return -1
    }
}
